File: programmers/2_주차_요금_계산.py
import math

def solution(fees, records):
    default, default_cost, unit, unit_cost = fees
    parking = {}
    for record in records:
        time, number, do = record.split()
        time = int(time[:2])*60 + int(time[3:])
        
        if not number in parking:
            parking[number] = [-1, 0]
        if do=='IN':
            parking[number][0] = time
        else:
            parking[number][1] += time - parking[number][0]
            parking[number][0] = -1
            
    answer = []
    for number in sorted(parking.keys()):
        if parking[number][0] != -1:
            parking[number][1] += 1439 - parking[number][0]
        
        p = parking[number]
        c = default_cost
        if p[1] > default:
            c += math.ceil((p[1] - default) / unit) * unit_cost
            
        answer.append(c)
        
    return answer

File: programmers/2_주차_요금_계산.js
function solution(fees, records) {
  const [dt, dp, ut, up] = fees;

  const park = [];
  const use = [];
  records.forEach((record) => {
    const [t, k, method] = record.split(" ");
    const [h, m] = t.split(":").map(Number);
    const time = h * 60 + m;
    const key = Number(k);

    if (use[key] === undefined) {
      use[key] = 0;
    }

    if (method === "IN") {
      park[key] = time;
    } else {
      use[key] += time - park[key];
      park[key] = 23 * 60 + 59;
    }
  });

  park.forEach((value, key) => {
    use[key] += 23 * 60 + 59 - park[key];
  });

  const price = {};
  use.forEach((value, key) => {
    price[key] = dp;

    if (value > dt) {
      price[key] += Math.ceil((value - dt) / ut) * up;
    }
  });

  return Object.entries(price)
    .sort(([ka, va], [kb, vb]) => ka - kb)
    .map(([k, v]) => v);
}
